### PR TITLE
AP_Arming: AP_Arming ensures mission items present when in auto mode

### PR DESCRIPTION
**Pull request by peterbarker**
Henry and I pair-programmed this.

We tested it in SITL (finding some other issues along the way), and also wrote some auto-tests for it.

### CODE REVIEW

**Review by Hwurzburg — APPROVED** [review]

File: libraries/AP_Arming/AP_Arming.cpp [hunk 1]
```diff
@@ -855,6 +856,15 @@ bool AP_Arming::mission_checks(bool report)
         check_failed(ARMING_CHECK_MISSION, report, "Failed to open %s", AP_MISSION_SDCARD_FILENAME);
     }
 #endif
+
+    // do not allow arming if there are no mission items and we are in
+    // (e.g.) AUTO mode
+    if (AP::vehicle()->current_mode_requires_mission() &&
```
Comment (IamPete1) [inline libraries/AP_Arming/AP_Arming.cpp:862]: Should this be before the per item checks? You could also break out the nullptr check from 803 to give that more useful message in that case.

File: libraries/AP_Arming/AP_Arming.cpp [hunk 2]
```diff
@@ -855,6 +856,15 @@ bool AP_Arming::mission_checks(bool report)
...
+        (mission == nullptr || mission->num_commands() <= 1)) {
+        check_failed(ARMING_CHECK_MISSION, report, "Mode requires non-empty mission");
```
Comment (IamPete1) [inline libraries/AP_Arming/AP_Arming.cpp:864]:
```suggestion
        check_failed(ARMING_CHECK_MISSION, report, "Mode requires mission");
```
I think "non-empty" is redundant.

**Review by tridge — APPROVED** [review]

File: ArduCopter/Copter.h [hunk 3]
```diff
@@ -851,6 +851,13 @@ class Copter : public AP_Vehicle {
     // called when an attempt to change into a mode is unsuccessful:
     void mode_change_failed(const Mode *mode, const char *reason);
     uint8_t get_mode() const override { return (uint8_t)flightmode->mode_number(); }
+    bool current_mode_requires_mission() const override {
+#if MODE_AUTO_ENABLED == ENABLED
```
Comment (rmackay9) [inline ArduCopter/Copter.h:855]: Looking good, could you move these lines to the .cpp file?

**Comment by rmackay9** [comment]
We can merge after the two small fixes requested..

**Comment by Hwurzburg** [comment]
@peterbarker please check my work...SITLs fine for Copter

**Comment by peterbarker** [comment]
> @peterbarker please check my work...SITLs fine for Copter

The cpp code you dropped within `#if AP_SCRIPTING_ENABLED`, so wasn't being included on 1MB boards.

I've fixed that and the commit message.

**Pull request merged**